ingestion/src/metadata/ingestion/source/api/rest/metadata.py

```diff
@@ -0,0 +1,238 @@
+    CreateAPICollectionRequest,
+)
+from metadata.generated.schema.api.data.createAPIEndpoint import (
+    CreateAPIEndpointRequest,
+)
+from metadata.generated.schema.entity.data.apiCollection import APICollection
+from metadata.generated.schema.entity.data.apiEndpoint import ApiRequestMethod
+from metadata.generated.schema.entity.services.connections.apiService.restConnection import (
+    RESTConnection,
+)
+from metadata.generated.schema.entity.services.ingestionPipelines.status import (
+    StackTraceError,
+)
+from metadata.generated.schema.metadataIngestion.workflow import (
+    Source as WorkflowSource,
+)
+from metadata.generated.schema.type.apiSchema import APISchema
+from metadata.generated.schema.type.basic import (
+    EntityName,
+    FullyQualifiedEntityName,
+    Markdown,
+)
+from metadata.generated.schema.type.schema import DataTypeTopic, FieldModel
+from metadata.ingestion.api.models import Either
+from metadata.ingestion.api.steps import InvalidSourceException
+from metadata.ingestion.ometa.ometa_api import OpenMetadata
+from metadata.ingestion.source.api.api_service import ApiServiceSource
+from metadata.utils import fqn
+from metadata.utils.logger import ingestion_logger
+
+logger = ingestion_logger()
+
+
+class RestSource(ApiServiceSource):
+    """
+    Source implementation to ingest REST data.
+
+    We will iterate on the registered collections, endpoints
+    and prepare an iterator of
+    """
+
+    def __init__(self, config: WorkflowSource, metadata: OpenMetadata):
+        super().__init__(config, metadata)
+
+    @classmethod
+    def create(
+        cls, config_dict, metadata: OpenMetadata, pipeline_name: Optional[str] = None
+    ):
+        config: WorkflowSource = WorkflowSource.model_validate(config_dict)
+        connection: RESTConnection = config.serviceConnection.root.config
+        if not isinstance(connection, RESTConnection):
+            raise InvalidSourceException(
+                f"Expected RESTConnection, but got {connection}"
+            )
+        return cls(config, metadata)
+
+    def get_api_collections(self, *args, **kwargs) -> Iterable[Any]:
+        """
+        Method to list all collections to process.
+        Here is where filtering happens
+        """
+        json_response = {}
+        try:
+            json_response = self.connection.json()
+        except Exception as err:
+            logger.error(f"Error while fetching collections from schema URL :{err}")
+
+        for collection in json_response.get("tags", []):
+            if not collection.get("name"):
+                continue
+            yield collection
+
+    def yield_api_collection(
+        self, collection: dict
+    ) -> Iterable[Either[CreateAPICollectionRequest]]:
+        """Method to return api collection Entities"""
+        try:
+            collection_request = CreateAPICollectionRequest(
+                name=EntityName(collection.get("name")),
```

Comment:
let's prepare pydantic models, this is error-prone:
1. `collection.get("name")` might be `None`
2. `EntityName` does not allow `None`
3. The error better tell us the data is not behaving properly, rather than enigmatic pydantic errors.

Let's properly parse the payload and validate early (shift left as many errors as possible), then let the rest of the flow be "dumb" on how to treat the results, instead of having to be aware all the time about the shape of our data